File: src/main/java/com/amarildo/TxtzExtractor.java
package com.amarildo;


import org.jetbrains.annotations.NotNull;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class TxtzExtractor {

    public static void run(@NotNull Path filePath) {
        Path cartellaParent = filePath.getParent();
        String txtzFilePath = String.valueOf(filePath);
        String destinationFolder = cartellaParent.toString();  // Inserisci qui il percorso di destinazione

        try {
            extractTxtzFile(txtzFilePath, destinationFolder);
            processExtractedFiles(txtzFilePath, destinationFolder);
            System.out.println("Operazione completata con successo.");
        } catch (IOException e) {
            e.printStackTrace();
            System.err.println("Errore durante l'operazione: " + e.getMessage());
        }
    }

    // Metodo per estrarre il contenuto di un file .txtz
    private static void extractTxtzFile(String txtzFilePath, String destinationFolder) throws IOException {
        try (ZipInputStream zipIn = new ZipInputStream(new FileInputStream(txtzFilePath))) {
            ZipEntry entry;
            while ((entry = zipIn.getNextEntry()) != null) {
                File outputFile = new File(destinationFolder, entry.getName());
                if (entry.isDirectory()) {
                    outputFile.mkdirs();
                } else {
                    outputFile.getParentFile().mkdirs();
                    try (BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(outputFile))) {
                        byte[] buffer = new byte[4096];
                        int bytesRead;
                        while ((bytesRead = zipIn.read(buffer)) != -1) {
                            bos.write(buffer, 0, bytesRead);
                        }
                    }
                }
                zipIn.closeEntry();
            }
        }
    }

    // Metodo per processare i file estratti
    private static void processExtractedFiles(String txtzFilePath, String destinationFolder) throws IOException {
        File extractedFolder = new File(destinationFolder);

        // Cancella il file metadata.opf
        File metadataFile = new File(extractedFolder, "metadata.opf");
        if (metadataFile.exists() && metadataFile.isFile()) {
            metadataFile.delete();
        }

        // Rinomina il file index.txt con il nome del file .txtz originale ma con estensione .md
        File indexFile = new File(extractedFolder, "index.txt");
        String originalFileName = Paths.get(txtzFilePath).getFileName().toString();
        String baseFileName = originalFileName.replaceFirst("\\.txtz$", "");
        String sanitizedFolderName = sanitizeFolderName(baseFileName);

        if (indexFile.exists() && indexFile.isFile()) {
            File renamedFile = new File(extractedFolder, sanitizedFolderName + ".md");
            indexFile.renameTo(renamedFile);

            // Rinomina la cartella images con il nome del file originale
            File imagesFolder = new File(extractedFolder, "images");
            File renamedImagesFolder = new File(extractedFolder, sanitizedFolderName);
            if (imagesFolder.exists() && imagesFolder.isDirectory()) {
                imagesFolder.renameTo(renamedImagesFolder);
            }

            // Crea un file di backup e aggiorna i riferimenti alle immagini
            File backupFile = new File(extractedFolder, sanitizedFolderName + "_backup.md");
            updateMarkdownImageReferences(renamedFile, backupFile, sanitizedFolderName);
        }
    }

    // Metodo per aggiornare i riferimenti alle immagini nel file markdown
    private static void updateMarkdownImageReferences(File originalFile, File backupFile, String newImagesFolderName) throws IOException {
        // Pattern regex per trovare i riferimenti alle immagini nella forma "](images/000XXX.png)"
        Pattern pattern = Pattern.compile("\\]\\(images/(\\S+?\\.[^\\)]+)\\)");

        try (
                BufferedReader reader = new BufferedReader(new FileReader(originalFile));
                BufferedWriter writer = new BufferedWriter(new FileWriter(backupFile))
        ) {
            String line;
            while ((line = reader.readLine()) != null) {
                // Aggiorna i riferimenti alle immagini sostituendo "images" con il nuovo nome della cartella
                String updatedLine = pattern.matcher(line).replaceAll("](" + newImagesFolderName + "/$1)");
                writer.write(updatedLine);
                writer.newLine();
            }
        }
    }

    // Metodo per normalizzare il nome della cartella
    @NotNull
    private static String sanitizeFolderName(@NotNull String folderName) {
        // Converti tutto in minuscolo, sostituisci spazi con underscore e rimuovi caratteri speciali
        return folderName.toLowerCase()
                .replaceAll("[^a-z0-9]", "_")  // Sostituisce ogni carattere non alfanumerico con underscore
                .replaceAll("_+", "_");       // Rimuove eventuali underscore ripetuti
    }
}

File: src/main/kotlin/com/amarildo/TxtzExtractor.kt
package com.amarildo

import java.io.*
import java.nio.file.Path
import java.nio.file.Paths
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

class TxtzExtractor {
    companion object {
        fun run(filePath: Path?) {
            if (filePath == null) {
                println("filePath cannot be null")
                return
            }

            val txtzFilePath = filePath.toString()
            val destinationFolder = filePath.parent.toString()

            try {
                extractTxtzFile(txtzFilePath, destinationFolder)
                processExtractedFiles(txtzFilePath, destinationFolder)
                println("Operation completed successfully.")
            } catch (e: IOException) {
                e.printStackTrace()
                System.err.println("Error during operation: ${e.message}")
            }
        }

        // Method to extract the contents of a .txtz file
        private fun extractTxtzFile(
            txtzFilePath: String,
            destinationFolder: String,
        ) {
            ZipInputStream(FileInputStream(txtzFilePath)).use { zipIn ->
                var entry: ZipEntry?
                while (zipIn.nextEntry.also { entry = it } != null) {
                    val outputFile = File(destinationFolder, entry!!.name)
                    if (entry!!.isDirectory) { // if the destination folder does not exist -> create it
                        outputFile.mkdirs()
                    } else {
                        outputFile.parentFile.mkdirs()
                        BufferedOutputStream(FileOutputStream(outputFile)).use { bos ->
                            val buffer = ByteArray(4096)
                            var bytesRead: Int
                            while (zipIn.read(buffer).also { bytesRead = it } != -1) {
                                bos.write(buffer, 0, bytesRead)
                            }
                        }
                    }
                    zipIn.closeEntry()
                }
            }
        }

        // Method to process the extracted files
        private fun processExtractedFiles(
            txtzFilePath: String,
            destinationFolder: String,
        ) {
            val extractedFolder = File(destinationFolder)

            // Delete the metadata.opf file
            val metadataFile = File(extractedFolder, "metadata.opf")
            if (metadataFile.exists() && metadataFile.isFile) {
                metadataFile.delete()
            }

            // Rename index.txt with the original .txtz file name but with .md extension
            val indexFile = File(extractedFolder, "index.txt")
            val originalFileName = Paths.get(txtzFilePath).fileName.toString()
            val baseFileName = originalFileName.replaceFirst("\\.txtz$".toRegex(), "")
            val sanitizedFolderName = sanitizeFolderName(baseFileName)

            if (indexFile.exists() && indexFile.isFile) {
                val renamedFile = File(extractedFolder, "$sanitizedFolderName.md")
                indexFile.renameTo(renamedFile)

                // Rename the images folder with the original file name
                val imagesFolder = File(extractedFolder, "images")
                val renamedImagesFolder = File(extractedFolder, sanitizedFolderName)
                if (imagesFolder.exists() && imagesFolder.isDirectory) {
                    imagesFolder.renameTo(renamedImagesFolder)
                }

                // Create a backup file and update image references
                val backupFile = File(extractedFolder, "${sanitizedFolderName}_optimized.md")
                updateMarkdownImageReferences(renamedFile, backupFile, sanitizedFolderName)
            }
        }

        // Method to update image references in the markdown file
        private fun updateMarkdownImageReferences(
            originalFile: File,
            backupFile: File,
            newImagesFolderName: String,
        ) {
            // Regex pattern to find image references in the form "](images/000XXX.png)"
            val pattern = Regex("\\]\\(images/(\\S+?\\.[^\\)]+)\\)")

            BufferedReader(FileReader(originalFile)).use { reader ->
                BufferedWriter(FileWriter(backupFile)).use { writer ->
                    reader.lineSequence().forEach { line ->
                        // Update image references replacing "images" with the new folder name
                        val updatedLine =
                            pattern
                                .replace(line, "]($newImagesFolderName/$1)")
                                .replace("\\(", "(") // remove '\' before curly
                                .replace("\\)", ")")
                                .replace("Figur", "figur") // lowercase Figura 1.XX
                        writer.write(updatedLine)
                        writer.newLine()
                    }
                }
            }
        }

        // Method to sanitize folder name
        private fun sanitizeFolderName(folderName: String): String {
            // Convert to lowercase, replace spaces with underscores, and remove special characters
            return folderName
                .lowercase()
                .replace("[^a-z0-9]".toRegex(), "_") // Replace each non-alphanumeric character with underscore
                .replace("_+".toRegex(), "_") // Remove repeated underscores
        }
    }
}
